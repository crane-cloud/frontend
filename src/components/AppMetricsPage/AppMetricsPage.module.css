.ContentSection {
  padding: 1rem;
  display: grid;
  gap: 1rem;
}

.MetricsCardsSection {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 1rem;
  padding-bottom: 1rem;
}

.CardSizeDimensions {
  min-width: 15rem;
  display: flex;
  /* we dont set values the height
       is a pre-requiste for the line
       component, try removing or editing
       it and you will see why it matters.
    */
  height: 18rem;
  color: initial;
}

.SummaryCardDimentions {
  min-width: 100%;
  display: flex;
}
.TopBarSection {
  background-color: black;
}
.Page {
  margin: 0;
  background: #f1f1f1;
  max-height: 100vh;
  display: grid;
  grid-template-rows: auto 1fr auto;
  /* overflow: hidden; */
}

.SideBarSection {
  overflow: hidden;
  overflow-y: scroll;
  height: calc(100vh - 5rem);
}

.SideBarSection::-webkit-scrollbar {
  display: none;
  min-width: fit-content;
}

.SideBarSection {
  -ms-overflow-style: none;
  scrollbar-width: none;
}
.MainContentSection {
  /* divided into two rows */
  display: grid;
  grid-template-rows: 80px 4fr;
  overflow: hidden;
}
.ContentSection {
  display: flex;
  flex-direction: column;
  overflow-x: hidden;
  overflow-y: auto;
}

.SummaryCardContainer {
  background-color: #fff;
  display: flex;
  flex-direction: column;
  border: 1px solid #cccccc;
}

.CardHeaderSection {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 1rem;
  border-bottom: 1px solid #cccccc;
  gap: 1rem;
  flex-grow: 1;
}

.CardTitle {
  width: 100%;
}
.InnerCard {
  display: grid;
  grid-template-columns: 7fr 1fr 7fr 1fr 7fr;
  grid-template-rows: auto;
  grid-gap: 1rem;
  overflow-x: auto;
}

.InnerCardSections {
  display: grid;
  grid-template-columns: auto;
  grid-template-rows: 1fr 1fr;
  gap: 0.2rem;
}

.CardBodySection {
  display: flex;
  justify-content: flex-start;
  flex-grow: 1;
  align-items: center;
  padding: 1rem 2rem;
  min-height: 10rem;
}

.SummaryCardContainer:hover {
  border: 1px solid var(--secondary-color);
}
.verticalLine {
  border-left: 1px solid #cccccc;
  height: 80%;
  width: 1px;;
  padding: 1rem;
}
.InnerTitlesStart {
  color: #000;
  font-weight: bold;
  text-align: start;
  padding: 1rem 1rem 0rem 0rem;
}
.InnerTitlesMiddle {
  color: #000;
  font-weight: bold;
  text-align: center;
  padding-top: 1rem;
}
.InnerTitlesEnd {
  color: #000;
  font-weight: bold;
  text-align: end;
  padding-top: 1rem;
}

.InnerContentGrid {
  display: grid;
  grid-template-columns: auto;
  grid-template-rows: 1fr 1fr;
  grid-gap: 0.3rem;
}

.InnerContentEnd {
  text-align: end;
}
.InnerContentStatus {
  display: flex;
  align-items: center;
  justify-content: center;
  column-gap: .5rem;
  color: #555;
  font-size: 0.9rem;
}
.InnerContentAge {
  text-align: center;
}
.InnerContentLinkText {
  white-space: nowrap;
  overflow: hidden;
  color: #008ac1;
  text-overflow: ellipsis;
}

.InnerContentLinkText a{
  color: inherit;
  text-decoration: none;
} 

.InnerContentLink {
  display: grid;
  grid-template-columns: auto 20%;
  gap: 0.5rem;
}

.Icons {
  display: grid;
  grid-auto-flow: column;
  gap: 0.5rem;
  cursor: pointer;
  align-items: center;
}

.CopierDiv{
  display: flex;
  justify-content: flex-end;
  align-items: flex-start;
}
