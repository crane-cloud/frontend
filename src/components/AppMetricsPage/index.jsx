import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import InformationBar from "../InformationBar";
import Header from "../Header";
import SideBar from "../SideBar";
import MetricsCard from "../MetricsCard";
import { ReactComponent as MetricIcon } from "../../assets/images/resource-icon.svg";
import styles from "./AppMetricsPage.module.css";
import LineChartComponent from "../LineChart";
import LogsFrame from "../LogsFrame";
import getAppLogs from "../../redux/actions/getAppLogs";
import getAppMemory, { clearAppMemory } from "../../redux/actions/appMemory";
import getAppCPU from "../../redux/actions/appCPU";
import {
  formatAppMemoryMetrics,
  formatAppCPUMetrics,
  formatAppNetworkMetrics,
} from "../../helpers/formatMetrics";
import getAppNetwork from "../../redux/actions/appNetwork";
import AppStatus from "../AppStatus";
import { ReactComponent as CopyText } from "../../assets/images/copytextblue.svg";
import { ReactComponent as Checked } from "../../assets/images/checked.svg";

class AppMetricsPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      urlChecked: false,
    };

    this.getAppMemoryMetrics = this.getAppMemoryMetrics.bind(this);
    this.getAppCPUMetrics = this.getAppCPUMetrics.bind(this);
    this.getAppNetworkMetrics = this.getAppNetworkMetrics.bind(this);
    this.copyUrl = this.copyUrl.bind(this);
  }

  getAppInfo(id) {
    const { apps } = this.props;
    const found = apps.apps.find((app) => app.id === id);
    const info = {
      name: found.name,
      status: found.app_running_status,
      url: found.url,
      age: found.age,
      alias: found.alias,
      image: found.image,
      port: found.port,
    };

    return info;
  }

  componentDidMount() {
    const {
      getAppLogs,
      getAppMemory,
      getAppCPU,
      getAppNetwork,
      match: { params },
    } = this.props;
    const { projectID, appID } = params;

    getAppLogs({ projectID, appID }, { timestamps: true });
    clearAppMemory();
    getAppMemory(projectID, appID, {});
    getAppCPU(projectID, appID, {});
    getAppNetwork(projectID, appID, {});
  }

  getAppMemoryMetrics() {
    const { appID } = this.props.match.params;
    const { appMemoryMetrics } = this.props;
    const results = formatAppMemoryMetrics(appID, appMemoryMetrics);
    return results;
  }

  copyUrl() {
    const { params } = this.props.match;
    const { appID } = params;
    const app = this.getAppInfo(appID);
    navigator.clipboard.writeText(app.url);
    this.setState({ urlChecked: true });
  }

  getAppCPUMetrics() {
    const { appID } = this.props.match.params;
    const { appCPUMetrics } = this.props;
    const results = formatAppCPUMetrics(appID, appCPUMetrics);
    return results;
  }

  getAppNetworkMetrics() {
    const { appID } = this.props.match.params;
    const { appNetworkMetrics } = this.props;
    const results = formatAppNetworkMetrics(appID, appNetworkMetrics);
    return results;
  }

  render() {
    const { params } = this.props.match;
    const { projectID, appID } = params;
    const { logs, retrieveingLogs } = this.props;
    const { urlChecked } = this.state;

    const formattedMemoryMetrics = this.getAppMemoryMetrics();
    const formattedCPUMetrics = this.getAppCPUMetrics();
    const formattedNetworkMetrics = this.getAppNetworkMetrics();
    const appInfo = this.getAppInfo(appID);

    return (
      <div className={styles.Page}>
        <div className={styles.TopBarSection}>
          <Header />
        </div>
        <div className="MainSection">
          <div className="SideBarSection">
            <SideBar
              name={appInfo.name}
              params={params}
              pageRoute={this.props.location.pathname}
              allMetricsLink={`/projects/${projectID}/apps/${appID}/dashboard`}
              cpuLink={`/projects/${projectID}/apps/${appID}/cpu/`}
              memoryLink={`/projects/${projectID}/apps/${appID}/memory/`}
              databaseLink={`/projects/${projectID}/databases`}
              networkLink={`/projects/${projectID}/apps/${appID}/network/`}
              appLogsLink={`/projects/${projectID}/apps/${appID}/logs/`}
            />
          </div>
          <div className="MainContentSection">
            <div className={styles.InformationBarSection}>
              <InformationBar header={appInfo.name} viewAppLink={appInfo.url} />
            </div>
            <div className={`${styles.ContentSection} SmallContainer`}>
              <div
                className={
                  styles.SummaryCardContainer +
                  " " +
                  styles.SummaryCardDimentions
                }
              >
                <div className={styles.CardHeaderSection}>
                  <div className={styles.CardTitle}>App Summary</div>
                </div>
                <div className={styles.CardBodySection}>
                  <div className={styles.InnerCard}>
                    <div className={styles.InnerCardSections}>
                      <div className={styles.InnerContentGrid}>
                        <div className={styles.InnerTitlesStart}>App Name</div>
                        <div className={styles.InnerContentName}>
                          {appInfo.name}
                        </div>
                      </div>
                      <div className={styles.InnerContentGrid}>
                        <div className={styles.InnerTitlesStart}>App Url</div>
                        <div className={styles.InnerContentLink}>
                          <div className={styles.InnerContentLinkText}>
                            {appInfo.url}
                          </div>
                          <div className={styles.CopierDiv}>
                            <div className={styles.Icons}>
                              <CopyText onClick={this.copyUrl} />
                              {urlChecked === true ? <Checked /> : null}
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <hr />
                    <div className={styles.InnerCardSections}>
                      <div className={styles.InnerContentGrid}>
                        <div className={styles.InnerTitlesMiddle}>
                          App Status
                        </div>
                        <div className={styles.InnerContentStatus}>
                          <AppStatus appStatus={appInfo.status} />
                          <div>
                            {appInfo.status === "running" ? "Ready" : "Down"}
                          </div>
                        </div>
                      </div>
                      <div className={styles.InnerContentGrid}>
                        <div className={styles.InnerTitlesMiddle}>
                          Date Deployed
                        </div>
                        <div className={styles.InnerContentAge}>
                          {appInfo.age}
                        </div>
                      </div>
                    </div>
                    <hr />
                    <div className={styles.InnerCardSections}>
                      <div className={styles.InnerContentGrid}>
                        <div className={styles.InnerTitlesEnd}>App Alias</div>
                        <div className={styles.InnerContentEnd}>
                          {appInfo.alias}
                        </div>
                      </div>
                      <div className={styles.InnerContentGrid}>
                        <div className={styles.InnerTitlesEnd}>Port</div>
                        <div className={styles.InnerContentEnd}>
                          {appInfo.port}
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div className={styles.MetricsCardsSection}>
                <Link
                  to={{
                    pathname: `/projects/${projectID}/apps/${appID}/memory/`,
                  }}
                >
                  <MetricsCard
                    icon={<MetricIcon />}
                    title="Memory"
                    className={styles.CardSizeDimensions}
                  >
                    <LineChartComponent
                      lineDataKey="memory"
                      preview
                      data={formattedMemoryMetrics}
                    />
                  </MetricsCard>
                </Link>
                <Link
                  to={{
                    pathname: `/projects/${projectID}/apps/${appID}/cpu/`,
                  }}
                >
                  <MetricsCard
                    icon={<MetricIcon />}
                    title="CPU"
                    className={styles.CardSizeDimensions}
                  >
                    <LineChartComponent
                      lineDataKey="cpu"
                      preview
                      data={formattedCPUMetrics}
                    />
                  </MetricsCard>
                </Link>
                <Link
                  to={{
                    pathname: `/projects/${projectID}/apps/${appID}/network/`,
                  }}
                >
                  <MetricsCard
                    icon={<MetricIcon />}
                    title="Network"
                    className={styles.CardSizeDimensions}
                  >
                    <LineChartComponent
                      lineDataKey="network"
                      preview
                      data={formattedNetworkMetrics}
                    />
                  </MetricsCard>
                </Link>
              </div>
              <div className={styles.LogsSection}>
                <Link
                  to={{
                    pathname: `/projects/${projectID}/apps/${appID}/logs/`,
                  }}
                >
                  <LogsFrame
                    loading={retrieveingLogs}
                    data={logs}
                    title={`${appInfo.name} logs`}
                  />
                </Link>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  const { logs, retrievedLogs, retrieveingLogs } = state.appLogsReducer;

  const { isFetchingAppMemory, appMemoryMetrics, appMemoryMessage } =
    state.appMemoryReducer;

  const { isFetchingCPU, appCPUMetrics, cpuMessage } = state.appCpuReducer;

  const { appNetworkMetrics, isFetchingAppNetwork, appNetworkMessage } =
    state.appNetworkReducer;

  const { apps } = state.appsListReducer;

  return {
    isFetchingAppMemory,
    appMemoryMetrics,
    appMemoryMessage,
    logs,
    retrievedLogs,
    retrieveingLogs,
    isFetchingCPU,
    appCPUMetrics,
    cpuMessage,
    appNetworkMetrics,
    isFetchingAppNetwork,
    appNetworkMessage,
    apps,
  };
};

const mapDispatchToProps = {
  getAppMemory,
  clearAppMemory,
  getAppLogs,
  getAppCPU,
  getAppNetwork,
};

AppMetricsPage.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      projectID: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  logs: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(AppMetricsPage);
