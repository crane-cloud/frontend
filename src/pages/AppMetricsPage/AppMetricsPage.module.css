.ContentSection {
  padding: 1rem;
  display: grid;
  gap: 1rem;
}

.MetricsCardsSection {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 1rem;
  padding-bottom: 1rem;
}

.CardSizeDimensions {
  min-width: 15rem;
  display: flex;
  /* we dont set values the height
       is a pre-requiste for the line
       component, try removing or editing
       it and you will see why it matters.
    */
  height: 18rem;
  color: initial;
}

.SummaryCardDimentions {
  min-width: 100%;
  display: flex;
  border-radius: 15px;
}

.SummarySectionArea {
  padding: 3rem 0 3rem 0;
}

.TopBarSection {
  background-color: black;
}

.Page {
  margin: 0;
  background: var(--gray-bg-light-color);
  max-height: 100vh;
  display: grid;
  grid-template-rows: auto 1fr auto;
  /* overflow: hidden; */
}

.SideBarSection {
  overflow: hidden;
  overflow-y: scroll;
  height: calc(100vh - 5rem);
}

.SideBarSection::-webkit-scrollbar {
  display: none;
  min-width: fit-content;
}

.SideBarSection {
  -ms-overflow-style: none;
  scrollbar-width: none;
}

.MainContentSection {
  /* divided into two rows */
  display: grid;
  grid-template-rows: 80px 4fr;
  overflow: hidden;
}

.ContentSection {
  display: flex;
  flex-direction: column;
  overflow-x: hidden;
  overflow-y: auto;
}

.SummaryCardContainer {
  background-color: #fff;
  display: flex;
  flex-direction: column;
  border: 1px solid #cccccc;
}

.CardHeaderSection {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 1rem;
  border-bottom: 1px solid #cccccc;
  gap: 1rem;
  flex-grow: 1;
}

.CardTitle {
  width: 100%;
}

.InnerCard {
  display: grid;
  grid-template-columns: 7fr 1fr 7fr 1fr 7fr;
  grid-template-rows: auto;
  grid-gap: 1rem;
  overflow-x: auto;
  width: -webkit-fill-available;
}
.NoteBookInnerCard {
  display: grid;
  grid-template-columns: 7fr 1fr 7fr ;
  grid-template-rows: auto;
  grid-gap: 1rem;
  overflow-x: auto;
  width: -webkit-fill-available;
}

.InnerCardSections {
  display: grid;
  grid-template-columns: auto;
  grid-template-rows: 1fr 1fr;
  gap: 0.2rem;
}

.InnerContentSection {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.InnerContentLeftSection {
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 1.5rem;
}

.CardBodySection {
  display: flex;
  justify-content: center;
  flex-grow: 1;
  align-items: center;
  padding: 1rem 2rem;
  min-height: 10rem;
}

.SummaryCardContainer:hover {
  border: 1px solid var(--secondary-color);
}

.verticalLine {
  border-left: 1px solid #cccccc;
  height: 80%;
  width: 1px;
  ;
  padding: 1rem;
}

.InnerTitlesStart {
  color: #000;
  font-weight: bold;
  text-align: start;
  padding: 1rem 1rem 0rem 0rem;
}

.InnerItemCopy:hover {
  cursor: pointer;
  color: var(--primary-color);
}

.InnerItemCopy .TooltipWordPlaceholder {
  font-weight: normal;
}

.InnerTitlesMiddle {
  color: #000;
  font-weight: bold;
  text-align: center;
  padding-top: 1rem;
}

.InnerTitlesEnd {
  color: #000;
  font-weight: bold;
  text-align: end;
  padding-top: 1rem;
}

.InnerContentGrid {
  display: grid;
  grid-template-columns: auto;
  grid-template-rows: 1fr 1fr;
  grid-gap: 0.3rem;
}

.InnerContentEnd {
  text-align: end;
}

.InnerContentStatus {
  display: flex;
  align-items: center;
  justify-content: center;
  column-gap: .5rem;
  color: #555;
  font-size: 0.9rem;
}

.InnerContentAge {
  text-align: center;
}

.InnerContentLinkText {
  white-space: nowrap;
  overflow: hidden;
  color: var(--primary-color);
  text-overflow: ellipsis;
}

.InnerContentLinkText a {
  color: inherit;
  text-decoration: none;
}

.InnerContentWarnText {
  overflow: hidden;
  color: #fb0c0c;
  text-overflow: ellipsis;
  cursor: pointer;
  text-decoration: underline;
}

.InnerContentLink {
  display: grid;
  grid-template-columns: auto 20%;
  gap: 0.5rem;
}

.Icons {
  display: grid;
  grid-auto-flow: column;
  gap: 0.5rem;
  cursor: pointer;
  align-items: center;
}

.CopierDiv {
  display: flex;
  justify-content: flex-end;
  align-items: flex-start;
}

.MetricCardsSection {
  display: grid;
  width: 100%;
  grid-template-columns: repeat(3, 1fr);
  gap: 1rem;
}

.Redeploy {
  display: flex;
  gap: 1rem;
  align-items: center;
  cursor: pointer;
}

.DeployText {
  color: #fb0c0c;
}

@media only screen and (max-width: 650px) {
  .MetricCardsSection {
    grid-template-columns: repeat(1, 1fr);
  }

  .ContentSection {
    padding-top: 2.5rem;
  }
}

@media only screen and (max-width: 900px) and (min-width: 481px) {
  .MetricCardsSection {
    grid-template-columns: repeat(2, 1fr);
  }
}