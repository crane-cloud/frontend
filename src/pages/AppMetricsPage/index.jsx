import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import Spinner from "../../components/Spinner";
import { Redirect } from "react-router-dom";
import MetricsCard from "../../components/MetricsCard";
import { ReactComponent as RedeployICon } from "../../assets/images/redeploy.svg";
import { ReactComponent as CPUIcon } from "../../assets/images/cpu.svg";
import { ReactComponent as NetworkIcon } from "../../assets/images/wifi.svg";
import { ReactComponent as MemoryIcon } from "../../assets/images/hard-drive.svg";

import styles from "./AppMetricsPage.module.css";
import LineChartComponent from "../../components/LineChart";
import LogsFrame from "../../components/LogsFrame";
import {
  handlePostRequestWithDataObject,
  handlePostRequestWithOutDataObject,
  handleGetRequest,
} from "../../apis/apis";
import getAppMemory, { clearAppMemory } from "../../redux/actions/appMemory";
import getAppCPU from "../../redux/actions/appCPU";
import {
  formatAppMemoryMetrics,
  formatAppCPUMetrics,
  formatAppNetworkMetrics,
} from "../../helpers/formatMetrics";
import revertUrl, { clearUrlRevertState } from "../../redux/actions/revertUrl";
import getAppNetwork from "../../redux/actions/appNetwork";
import AppStatus from "../../components/AppStatus";
import { ReactComponent as CopyText } from "../../assets/images/copytextblue.svg";
import { ReactComponent as Checked } from "../../assets/images/checked.svg";
import DashboardLayout from "../../components/Layouts/DashboardLayout";

class AppMetricsPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      urlChecked: false,
      app: [],
      logs: [],
      fetchingLogs: true,
      logsError: "",
      redeploy: false,
      deployError: "",
      spin: false,
    };

    this.getAppMemoryMetrics = this.getAppMemoryMetrics.bind(this);
    this.getAppCPUMetrics = this.getAppCPUMetrics.bind(this);
    this.getAppNetworkMetrics = this.getAppNetworkMetrics.bind(this);
    this.regenerate = this.regenerate.bind(this);
    this.copyUrl = this.copyUrl.bind(this);
    this.renderRedirect = this.renderRedirect.bind(this);
  }

  getAppInfo() {
    const { app } = this.state;

    if (!app) {
      return null;
    }

    const info = {
      name: app.name,
      status: app.app_running_status,
      url: app.url,
      age: app.age,
      alias: app.alias,
      image: app.image,
      port: app.port,
      disable: app.disabled,
    };

    return info;
  }

  componentDidMount() {
    this.setState({ spin: true });
    const {
      getAppMemory,
      getAppCPU,
      getAppNetwork,
      clearUrlRevertState,
      match: { params },
    } = this.props;
    const { projectID, appID } = params;
    handlePostRequestWithDataObject(
      { timestamps: true },
      `/projects/${projectID}/apps/${appID}/logs`
    )
      .then((response) => {
        this.setState({
          logs: response.data.data.pods_logs,
          fetchingLogs: false,
        });
      })
      .catch((error) => {
        this.setState({
          logsError: "Failed to fetch logs",
          fetchingLogs: false,
        });
      });

    handleGetRequest(`/apps/${appID}`)
      .then((response) => {
        this.setState({
          app: response.data.data.apps,
          spin: false,
        });
      })
      .catch((error) => {
        this.setState({
          redeploy: true,
          deployError: error?.response?.data?.message,
        });
      });

    clearAppMemory();
    clearUrlRevertState();
    getAppMemory(projectID, appID, {});
    getAppCPU(projectID, appID, {});
    getAppNetwork(projectID, appID, {});
  }

  getAppMemoryMetrics() {
    const { appID } = this.props.match.params;
    const { appMemoryMetrics } = this.props;
    const results = formatAppMemoryMetrics(appID, appMemoryMetrics);
    return results;
  }
  // "Application does not exist on the cluster"
  copyUrl() {
    const app = this.getAppInfo();
    navigator.clipboard.writeText(app.url);
    this.setState({ urlChecked: true });
  }

  getAppCPUMetrics() {
    const { appID } = this.props.match.params;
    const { appCPUMetrics } = this.props;
    const results = formatAppCPUMetrics(appID, appCPUMetrics);
    return results;
  }

  getAppNetworkMetrics() {
    const { appID } = this.props.match.params;
    const { appNetworkMetrics } = this.props;
    const results = formatAppNetworkMetrics(appID, appNetworkMetrics);
    return results;
  }
  regenerate() {
    const {
      revertUrl,
      match: { params },
    } = this.props;
    revertUrl(params.appID);
  }
  renderRedirect = () => {
    const { isReverted, clearUrlRevertState } = this.props;
    if (isReverted) {
      clearUrlRevertState();
      return <Redirect to={`/projects`} noThrow />;
    }
  };

  handleRedeployApp = () => {
    this.setState({ spin: true });
    const {
      match: { params },
    } = this.props;

    const { appID } = params;

    handlePostRequestWithOutDataObject({}, `/apps/${appID}/redeploy`)
      .then((response) => {
        window.location.reload();
      })
      .catch((error) => {
        console.error("Failed to redeploy app:", error);
      });
  };

  render() {
    const { params } = this.props.match;
    const { projectID, appID } = params;
    // eslint-disable-next-line no-unused-vars
    const { isReverting, isReverted } = this.props;
    const { logs, logsError, fetchingLogs, urlChecked, redeploy, spin } =
      this.state;

    const formattedMemoryMetrics = this.getAppMemoryMetrics();
    const formattedCPUMetrics = this.getAppCPUMetrics();
    const formattedNetworkMetrics = this.getAppNetworkMetrics();
    const appInfo = this.getAppInfo();
    return (
      <DashboardLayout
        name={appInfo.name}
        header={appInfo.name}
        viewAppLink={appInfo.url}
      >
        <div
          className={
            styles.SummaryCardContainer + " " + styles.SummaryCardDimentions
          }
        >
          <div className={styles.CardHeaderSection}>
            <div className={styles.CardTitle}>App Summary</div>
          </div>
          {spin ? (
            <div className={styles.SummarySectionArea}>
              <div className="SpinnerWrapper">
                <Spinner size="big" />
              </div>
            </div>
          ) : (
            <div className={styles.CardBodySection}>
              <div className={styles.InnerCard}>
                <div className={styles.InnerCardSections}>
                  <div className={styles.InnerContentGrid}>
                    <div className={styles.InnerTitlesStart}>App Name</div>
                    <div className={styles.InnerContentName}>
                      {appInfo.name}
                    </div>
                  </div>
                  <div className={styles.InnerContentGrid}>
                    <div className={styles.InnerTitlesStart}>App Url</div>
                    {appInfo.url ? (
                      <>
                        {redeploy ? (
                          <div
                            className={styles.Redeploy}
                            onClick={this.handleRedeployApp}
                          >
                            <RedeployICon />
                            <span className={styles.DeployText}>
                              Redeploy App
                            </span>
                          </div>
                        ) : (
                          <div className={styles.InnerContentLink}>
                            <div className={styles.InnerContentLinkText}>
                              <a
                                href={appInfo.url}
                                rel="noopener noreferrer"
                                target="_blank"
                              >
                                {appInfo.url}
                              </a>
                            </div>
                            <div className={styles.CopierDiv}>
                              <div className={styles.Icons}>
                                <CopyText onClick={this.copyUrl} />
                                {urlChecked === true ? <Checked /> : null}
                              </div>
                            </div>
                          </div>
                        )}
                      </>
                    ) : (
                      <>
                        {isReverting ? (
                          <Spinner />
                        ) : (
                          <div
                            className={styles.InnerContentWarnText}
                            onClick={() => {
                              this.regenerate();
                            }}
                          >
                            Click to re-generate url
                          </div>
                        )}
                      </>
                    )}
                  </div>
                </div>

                <hr />
                <div className={styles.InnerCardSections}>
                  <div className={styles.InnerContentGrid}>
                    <div className={styles.InnerTitlesMiddle}>App Status</div>
                    <div className={styles.InnerContentStatus}>
                      <AppStatus appStatus={appInfo.status} />
                      <div>
                        {appInfo.status === "disabled" ? (
                          <div className={styles.DeployText}>Disabled</div>
                        ) : appInfo.status === "running" ? (
                          "Ready"
                        ) : (
                          "Down"
                        )}
                      </div>
                    </div>
                  </div>
                  <div className={styles.InnerContentGrid}>
                    <div className={styles.InnerTitlesMiddle}>
                      Date Deployed
                    </div>
                    <div className={styles.InnerContentAge}>{appInfo.age}</div>
                  </div>
                </div>
                <hr />
                <div className={styles.InnerCardSections}>
                  <div className={styles.InnerContentGrid}>
                    <div className={styles.InnerTitlesEnd}>App Alias</div>
                    <div className={styles.InnerContentEnd}>
                      {appInfo.alias}
                    </div>
                  </div>
                  <div className={styles.InnerContentGrid}>
                    <div className={styles.InnerTitlesEnd}>Port</div>
                    <div className={styles.InnerContentEnd}>{appInfo.port}</div>
                  </div>
                </div>
              </div>
            </div>
          )}
        </div>
        <div className={styles.MetricCardsSection}>
          <Link
            to={{
              pathname: `/projects/${projectID}/apps/${appID}/memory/`,
            }}
          >
            <MetricsCard
              icon={<MemoryIcon />}
              title="Memory"
              className="CardDimensions"
            >
              <LineChartComponent
                lineDataKey="memory"
                preview
                data={formattedMemoryMetrics}
              />
            </MetricsCard>
          </Link>
          <Link
            to={{
              pathname: `/projects/${projectID}/apps/${appID}/cpu/`,
            }}
          >
            <MetricsCard
              icon={<CPUIcon />}
              title="CPU"
              className="CardDimensions"
            >
              <LineChartComponent
                lineDataKey="cpu"
                preview
                data={formattedCPUMetrics}
              />
            </MetricsCard>
          </Link>
          <Link
            to={{
              pathname: `/projects/${projectID}/apps/${appID}/network/`,
            }}
          >
            <MetricsCard
              icon={<NetworkIcon />}
              title="Network"
              className="CardDimensions"
            >
              <LineChartComponent
                lineDataKey="network"
                preview
                data={formattedNetworkMetrics}
              />
            </MetricsCard>
          </Link>
        </div>
        <div className={styles.LogsSection}>
          <Link
            to={{
              pathname: `/projects/${projectID}/apps/${appID}/logs/`,
            }}
          >
            <LogsFrame
              loading={fetchingLogs}
              data={logs}
              title={`${appInfo.name} logs`}
              error={logsError}
            />
          </Link>
        </div>
      </DashboardLayout>
    );
  }
}

export const mapStateToProps = (state) => {
  const { isFetchingAppMemory, appMemoryMetrics, appMemoryMessage } =
    state.appMemoryReducer;
  const { isFetchingCPU, appCPUMetrics, cpuMessage } = state.appCpuReducer;
  const { appNetworkMetrics, isFetchingAppNetwork, appNetworkMessage } =
    state.appNetworkReducer;
  const { apps } = state.appsListReducer;
  const { isReverting, isReverted } = state.revertUrlReducer;

  return {
    isFetchingAppMemory,
    appMemoryMetrics,
    appMemoryMessage,
    isFetchingCPU,
    appCPUMetrics,
    cpuMessage,
    appNetworkMetrics,
    isFetchingAppNetwork,
    appNetworkMessage,
    apps,
    isReverted,
    isReverting,
  };
};

const mapDispatchToProps = {
  getAppMemory,
  clearAppMemory,
  getAppCPU,
  getAppNetwork,
  clearUrlRevertState,
  revertUrl,
};

AppMetricsPage.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      projectID: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(AppMetricsPage);
