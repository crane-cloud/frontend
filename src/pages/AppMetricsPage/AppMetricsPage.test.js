/* eslint-disable no-undef */
import React from "react";
import { render } from "@testing-library/react";

import AppMetricsPage, { mapStateToProps } from ".";

const AppMetricsPageProps = {
  logs: ["log1", "log2"],
  match: { params: { projectID: "1" } },
  getAppMemory: jest.fn(),
  getAppCPU: jest.fn(),
  getAppNetwork: jest.fn(),
  clearUrlRevertState: jest.fn(),
};
// {} ={}
describe("Testing the App Metrics Page component", () => {
  const WrapperAppMetricsPage = AppMetricsPage.WrappedComponent;
  const AppMetricsPageComponent = render(
    <WrapperAppMetricsPage {...AppMetricsPageProps} />
  );
  it("should match the snapshot for AppMetricsPage after adding props", () => {
    AppMetricsPageComponent.setProps(AppMetricsPageProps);
    expect(AppMetricsPageComponent).toBeDefined();
  });
  it("matchs the AppMetricsPage component snapshot", () => {
    expect(AppMetricsPageComponent).toMatchSnapshot();
  });
});

describe("Testing the exported mapstate to props and dispatch", () => {
  it("matches the Appmetricspage mapstostate", () => {
    expect(
      mapStateToProps({
        appMemoryReducer: {
          isFetchingAppMemory: false,
          appMemoryMetrics: [],
          appMemoryMessage: "",
        },
        appCpuReducer: {
          isFetchingCPU: false,
          appCPUMetrics: [],
          cpuMessage: "",
        },
        appNetworkReducer: {
          appNetworkMetrics: [],
          isFetchingAppNetwork: false,
          appNetworkMessage: "",
        },
        appsListReducer: { apps: [] },
        revertUrlReducer: {
          isReverting: false,
          isReverted: false,
        },
      })
    ).toEqual({
      isFetchingAppMemory: false,
      appMemoryMetrics: [],
      appMemoryMessage: "",
      isFetchingCPU: false,
      appCPUMetrics: [],
      cpuMessage: "",
      appNetworkMetrics: [],
      isFetchingAppNetwork: false,
      appNetworkMessage: "",
      apps: [],
      isReverted: false,
      isReverting: false,
    });
  });
});
