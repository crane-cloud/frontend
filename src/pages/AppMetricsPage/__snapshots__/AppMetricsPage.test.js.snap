// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the App Metrics Page component matches the ProjectDashboardPage component snapshot 1`] = `ShallowWrapper {}`;
